frameless-runtime/src/utxo.rs

```diff
@@ -86,25 +110,91 @@ pub trait UtxoSet {
     /// - Not consuming it but marking it as spent
     ///
     fn nullify(utxo_ref: UtxoRef) -> Option<Utxo>;
+
+    fn peak(utxo_ref: UtxoRef) -> Option<Utxo> {
+        let encoded_utxo = sp_io::storage::get(&utxo_ref.encode())?;
+        match Utxo::decode(&mut &encoded_utxo[..]) {
+            Ok(utxo) => Some(utxo),
+            Err(_) => None,
+        }
+    }
+}
+
+pub trait Get<T> {
+    fn get(&self) -> T;
 }
 
 /// The API of a Tuxedo Piece
 pub trait TuxedoPiece {
     /// The type of data stored in Outputs associated with this Piece
     type Data: Encode + Decode;
+    const TYPE_ID: TypeId;
 
     /// The validation function to determine whether a given input can be consumed.
     fn validate(transaction: Transaction) -> bool;
 }
 
+pub struct PieceExtracter<Piece>(PhantomData<Piece>);
+impl<Piece: TuxedoPiece> PieceExtracter<Piece> {
+    pub fn extract(key: UtxoRef) -> Result<Piece::Data, ()> {
+        let encoded_utxo = sp_io::storage::get(&key.encode()).ok_or(())?;
+        let utxo = Utxo::decode(&mut &encoded_utxo[..]).map_err(|_| ())?;
+        if utxo.data_id != Piece::TYPE_ID {
+            return Err(())
+        }
+        let piece_data = Piece::Data::decode(&mut &utxo.data[..]).map_err(|_| ())?;
+        Ok(piece_data)
+    }
+}
+
 // User defined logic below for the STF..
 
 pub struct MoneyPiece; // Decodes Value -> u128
+impl UtxoSet for MoneyPiece {
+    fn contains(utxo_ref: UtxoRef) -> bool {
+        sp_io::storage::exists(&utxo_ref.encode())
+    }
+
+    fn insert(utxo_ref: UtxoRef, utxo: Utxo) -> bool {
+        sp_io::storage::set(&utxo_ref.encode(), &utxo.encode());
+        true
+    }
+
+    /// For the Money UTXO we just want to consume it.
+    fn nullify(utxo_ref: UtxoRef) -> Option<Utxo> {
+        let encoded_utxo = sp_io::storage::get(&utxo_ref.encode())?;
+        sp_io::storage::clear(&utxo_ref.encode());
+        match Utxo::decode(&mut &encoded_utxo[..]) {
+            Ok(utxo) => Some(utxo),
+            Err(_) => None,
+        }
+    }
+}
+
 impl TuxedoPiece for MoneyPiece {
     type Data = u128;
+    const TYPE_ID: TypeId = *b"1111";
 
     fn validate(transaction: Transaction) -> bool {
-        // decode the transaction output data as type `Data`
+        // Check that the input is unique and a set
+        {
+            let input_set: BTreeSet<_> = transaction.inputs.iter().collect();
+            if input_set.len() < transaction.inputs.len() {
+                return false;
+            }
+        }
+
+        // 1.) Check that the outputs referenced by each input can be redeemed
```

Comment:
This is the biggest disadvantage I see with the way your Call enum is structured. Having the utxo-style transaction embedded in each call variant means that each tuxedo piece will have to include this logic. We, the tuxedo core developers, could make this a little nicer for piece developers by providing a standard helper function that they can just call At least then they don't have to re-write it or copy paste it, but there is still the possibility that they forget to call it entirely.

Reply:
You are right can probably wrap it up nicely in a function or somehow make it default always done no matter what. I just see a path ahead and was going with it. I can swap to another paradigm if we want.